Native image: SoftDeleteStateManagement.INSTANCE field not registered for reflection

### Describe the bug

After #54083 bumped Hibernate ORM to 7.4.0.Final, native image builds fail at runtime when Hibernate tries to reflectively access `SoftDeleteStateManagement.INSTANCE`.

The new `Stateful.getStateManagement()` method (introduced in Hibernate ORM 7.4, `Stateful.java:53`) uses `getDeclaredField("INSTANCE")` to obtain the singleton instance. In native mode, this field is not registered for reflection, causing a `NoSuchFieldException`.

### Expected behavior

Native applications using Hibernate ORM should start successfully.

### Actual behavior

Application fails at startup with:

```
jakarta.persistence.PersistenceException: [PersistenceUnit: named-2] Unable to build Hibernate SessionFactory
  Caused by: org.hibernate.InstantiationException: Could not create StateManagement
    'org.hibernate.persister.state.internal.SoftDeleteStateManagement' due to: INSTANCE
  Caused by: java.lang.NoSuchFieldException: INSTANCE
    at java.base/java.lang.Class.getDeclaredField(DynamicHub.java:1478)
    at org.hibernate.mapping.Stateful.getStateManagement(Stateful.java:53)
```

### How to Reproduce?

Run the `hibernate-orm` module tests from quarkus-test-suite in native mode:

CI failures:
- https://github.com/quarkus-qe/quarkus-test-suite/actions/runs/27271567503/job/80546009359 (Linux x86_64 Native)
- https://github.com/quarkus-qe/quarkus-test-suite/actions/runs/27271567503/job/80546009468 (Linux AArch64 Native)

Failed tests: `MySqlHibernateSpatialIT`, `PostgisHibernateSpatialIT`, `ProdHibernateIT`

### Analysis

`SoftDeleteStateManagement` is a new class in Hibernate ORM 7.4 (by Gavin King). It follows the singleton pattern with a `public static final INSTANCE` field. The `Stateful.getStateManagement()` default method accesses it reflectively via `getDeclaredField("INSTANCE")`. The Quarkus Hibernate ORM extension's native image reflection configuration needs to register this field.

Other `StateManagement` implementations using the same pattern (e.g. `StandardStateManagement`) may also need registration.

### Environment info

- Quarkus: main (999-SNAPSHOT, includes #54083)
- Hibernate ORM: 7.4.0.Final
- GraalVM: Mandrel 25.0.3.0, JDK 25
- Architectures: both x86_64 and AArch64

--- a/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java
+++ b/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java
@@ -148,6 +148,7 @@ import io.quarkus.hibernate.orm.runtime.config.DialectVersions;
 import io.quarkus.hibernate.orm.runtime.customized.FormatMapperKind;
 import io.quarkus.hibernate.orm.runtime.customized.JsonFormatterCustomizationCheck;
 import io.quarkus.hibernate.orm.runtime.graal.RegisterServicesForReflectionFeature;
+import io.quarkus.hibernate.orm.runtime.graal.RegisterStateManagementForReflectionFeature;
 import io.quarkus.hibernate.orm.runtime.integration.HibernateOrmIntegrationStaticDescriptor;
 import io.quarkus.hibernate.orm.runtime.migration.MultiTenancyStrategy;
 import io.quarkus.hibernate.orm.runtime.proxies.PreGeneratedProxies;
@@ -211,6 +212,11 @@ public final class HibernateOrmProcessor {
         }
 
         return new NativeImageFeatureBuildItem(RegisterServicesForReflectionFeature.class);
+    }
+
+    @BuildStep
+    NativeImageFeatureBuildItem registerStateManagementForReflection() {
+        return new NativeImageFeatureBuildItem(RegisterStateManagementForReflectionFeature.class);
     }
 
     @BuildStep

--- a/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/softdelete/SoftDeleteEntity.java
+++ b/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/softdelete/SoftDeleteEntity.java
@@ -1,0 +1,41 @@
+package io.quarkus.hibernate.orm.softdelete;
+
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.Id;
+
+import org.hibernate.annotations.SoftDelete;
+
+@SoftDelete
+@Entity
+public class SoftDeleteEntity {
+
+    @Id
+    @GeneratedValue
+    Long id;
+
+    String name;
+
+    public SoftDeleteEntity() {
+    }
+
+    public SoftDeleteEntity(String name) {
+        this.name = name;
+    }
+
+    public Long getId() {
+        return id;
+    }
+
+    public void setId(Long id) {
+        this.id = id;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+}

--- a/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/softdelete/SoftDeleteTest.java
+++ b/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/softdelete/SoftDeleteTest.java
@@ -1,0 +1,61 @@
+package io.quarkus.hibernate.orm.softdelete;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import jakarta.inject.Inject;
+import jakarta.transaction.Transactional;
+
+import org.hibernate.Session;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusExtensionTest;
+
+/**
+ * Verifies that entities annotated with {@code @SoftDelete} work correctly.
+ * <p>
+ * This exercises the {@code Stateful.getStateManagement()} code path which
+ * reflectively accesses the {@code INSTANCE} field on {@code StateManagement}
+ * implementations via {@code getDeclaredField("INSTANCE")}. In native mode,
+ * these fields must be registered for reflection.
+ *
+ * @see <a href="https://github.com/quarkusio/quarkus/issues/54777">GitHub issue #54777</a>
+ */
+public class SoftDeleteTest {
+
+    @RegisterExtension
+    static QuarkusExtensionTest runner = new QuarkusExtensionTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClass(SoftDeleteEntity.class)
+                    .addAsResource("application.properties"));
+
+    @Inject
+    Session session;
+
+    @Test
+    @Transactional
+    public void testSoftDeleteEntityPersistAndFind() {
+        SoftDeleteEntity entity = new SoftDeleteEntity("test");
+        session.persist(entity);
+        session.flush();
+
+        SoftDeleteEntity found = session.find(SoftDeleteEntity.class, entity.getId());
+        assertThat(found).isNotNull();
+        assertThat(found.getName()).isEqualTo("test");
+    }
+
+    @Test
+    @Transactional
+    public void testSoftDeleteEntityRemove() {
+        SoftDeleteEntity entity = new SoftDeleteEntity("to-delete");
+        session.persist(entity);
+        session.flush();
+
+        Long id = entity.getId();
+        session.remove(entity);
+        session.flush();
+
+        SoftDeleteEntity found = session.find(SoftDeleteEntity.class, id);
+        assertThat(found).isNull();
+    }
+}

--- a/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/graal/RegisterStateManagementForReflectionFeature.java
+++ b/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/graal/RegisterStateManagementForReflectionFeature.java
@@ -1,0 +1,41 @@
+package io.quarkus.hibernate.orm.runtime.graal;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import org.graalvm.nativeimage.hosted.Feature;
+import org.graalvm.nativeimage.hosted.RuntimeReflection;
+import org.hibernate.persister.state.spi.StateManagement;
+
+/**
+ * Makes fields of reachable Hibernate StateManagement implementations accessible through
+ * {@link Class#getDeclaredField(String)}.
+ * <p>
+ * Hibernate ORM 7.4+ uses reflection to access the static {@code INSTANCE} field of
+ * StateManagement implementations at runtime. This feature ensures those fields are
+ * available for reflection in native mode.
+ * <p>
+ * See <a href="https://github.com/quarkusio/quarkus/issues/54777">GitHub issue #54777</a>.
+ */
+public class RegisterStateManagementForReflectionFeature implements Feature {
+
+    @Override
+    public String getDescription() {
+        return "Makes fields of reachable Hibernate StateManagement implementations accessible through Class#getDeclaredField()";
+    }
+
+    // The {@code duringAnalysis} method is invoked multiple times and increases the set of reachable types, thus we
+    // need to invoke {@link DuringAnalysisAccess#requireAnalysisIteration()} each time we register new fields.
+    private static final int ANTICIPATED_STATE_MANAGEMENT_CLASSES = 10;
+    private static final Set<Class<?>> registeredClasses = new HashSet<>(ANTICIPATED_STATE_MANAGEMENT_CLASSES);
+
+    @Override
+    public void duringAnalysis(DuringAnalysisAccess access) {
+        for (Class<?> stateManagement : access.reachableSubtypes(StateManagement.class)) {
+            if (registeredClasses.add(stateManagement)) {
+                RuntimeReflection.registerAllDeclaredFields(stateManagement);
+                access.requireAnalysisIteration();
+            }
+        }
+    }
+}

